src/keystore.c

```diff
@@ -30,6 +30,71 @@
 
 DPS_DEBUG_CONTROL(DPS_DEBUG_ON);
 
+DPS_KeyStore* DPS_CreateKeyStore(DPS_KeyAndIdentityHandler keyAndIdentityHandler, DPS_KeyHandler keyHandler,
+                                 DPS_CAHandler caHandler, DPS_CertHandler certHandler)
+{
+    DPS_DBGTRACE();
+
+    DPS_KeyStore* keyStore = calloc(1, sizeof(DPS_KeyStore));
+    if (keyStore) {
+        keyStore->keyAndIdentityHandler = keyAndIdentityHandler;
+        keyStore->keyHandler = keyHandler;
+        keyStore->caHandler = caHandler;
+        keyStore->certHandler = certHandler;
+    }
+    return keyStore;
+}
+
+void DPS_DestroyKeyStore(DPS_KeyStore* keyStore)
+{
+    DPS_DBGTRACE();
+
+    if (!keyStore) {
+        return;
+    }
+    free(keyStore);
+}
+
+DPS_Status DPS_SetKeyStoreData(DPS_KeyStore* keyStore, void* data)
+{
+    if (keyStore) {
+        keyStore->userData = data;
+        return DPS_OK;
+    } else {
+        return DPS_ERR_NULL;
+    }
+}
+
+void* DPS_GetKeyStoreData(const DPS_KeyStore* keyStore)
+{
+    return keyStore ? keyStore->userData : NULL;
+}
+
+DPS_KeyStore* DPS_KeyStoreHandle(DPS_KeyStoreRequest* request)
+{
+    return request ? request->keyStore : NULL;
+}
+
+DPS_Status DPS_SetKeyAndIdentity(DPS_KeyStoreRequest* request, const unsigned char* key, size_t keyLen, const unsigned char* id, size_t idLen)
+{
+    return request->setKeyAndIdentity ? request->setKeyAndIdentity(request, key, keyLen, id, idLen) : DPS_ERR_MISSING;
```

Comment:
Since you are setting the key here should you also call the `request->setKey(...)` callback?

Reply:
Under the hood there are two different functions used in mbedtls to handle these cases.  Rather than hold onto data between two separate calls (e.g. DPS_SetKeyIdentity and DPS_SetKey) and then make the underlying call, I decided to map 1:1 onto the mbedtls calls.

I may change this in the future to what you suggest, but for now this was the most straightforward path.